### PR TITLE
added STUN-based TIER1 self-discovery

### PR DESCRIPTION
**Pull request by pompon0**
If no public_addrs are configured, a validator node will try to query STUN servers to discover its own IP.
By default a list of public STUN servers provided by Google is used, bu the user is recommended to overwrite it with whatever STUN servers they trust OR, preferably, specify the proxy addresses manually.

### CODE REVIEW

File: chain/network/src/stun/mod.rs [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+/// Address of the format "<domain/ip>:<port>" of STUN servers.
+// TODO(gprusak): turn into a proper struct implementing Display and FromStr.
+pub type ServerAddr = String;
+
+pub(crate) type Error = stun::Error;
+
+const QUERY_TIMEOUT: time::Duration = time::Duration::seconds(5);
+
+pub(crate) async fn query(
```
Comment (mm-near) [inline chain/network/src/stun/mod.rs:19]: comment

File: chain/network/src/peer_manager/network_state/tier1.rs [hunk 2]
```diff
@@ -92,21 +93,45 @@ impl super::NetworkState {
...
+                // Query all the STUN servers in parallel.
+                let queries = stun_servers.iter().map(|addr| {
+                    let clock = clock.clone();
+                    let addr = addr.clone();
+                    self.spawn(async move { stun::query(&clock, &addr).await })
+                });
+                let mut node_ips = vec![];
+                for q in queries {
+                    match q.await.unwrap() {
+                        Err(err) => tracing::warn!(target:"network", "STUN lookup failed: {err}"),
```
Comment (mm-near) [inline chain/network/src/peer_manager/network_state/tier1.rs:118]: nit: Err should contain the information on which stun server was used.

File: chain/network/src/peer_manager/network_state/tier1.rs [hunk 3]
```diff
@@ -92,21 +93,45 @@ impl super::NetworkState {
...
+                        Ok(ip) => node_ips.push(ip),
+                    }
+                }
+                // Check that we have received non-zero responses and that they are consistent.
+                if node_ips.len() == 0 {
+                    vec![]
```
Comment (mm-near) [inline chain/network/src/peer_manager/network_state/tier1.rs:124]: this means that we cannot get the IP address, right? 

Is it visible somewhere?

I wonder if we should somehow export the current ip (in prometheus?)

File: chain/network/src/config_json.rs [hunk 4]
```diff
@@ -161,9 +177,8 @@ pub struct Config {
     /// Used only if this node is a validator and public_ips is empty (see
     /// description of public_ips field).  Format `<domain/ip>:<port>`, for
     /// example `stun.l.google.com:19302`.
-    // TODO: unskip, once the functionality is implemented.
-    #[serde(skip)] // TODO: add a default list.
-    pub trusted_stun_servers: Vec<String>,
+    #[serde(default = "default_trusted_stun_servers")]
```
Comment (mm-near) [inline chain/network/src/config_json.rs:180]: let's also mention it in changelog.md

File: chain/network/src/config_json.rs [hunk 5]
```diff
@@ -161,9 +177,8 @@ pub struct Config {
...
+    pub trusted_stun_servers: Vec<stun::ServerAddr>,
```
Comment (mm-near) [inline chain/network/src/config_json.rs:181]: I'd also mention that we'll query all of them in parallel (and not proceed if any of them returns a different ip than the rest)

**Review by mm-near — APPROVED** [review]

File: chain/network/src/stun/mod.rs (hunk 1 above)
Comment (pompon0) [reply chain/network/src/stun/mod.rs:19]: done

File: chain/network/src/peer_manager/network_state/tier1.rs (hunk 2 above)
Comment (pompon0) [reply chain/network/src/peer_manager/network_state/tier1.rs:118]: done

File: chain/network/src/peer_manager/network_state/tier1.rs (hunk 3 above)
Comment (pompon0) [reply chain/network/src/peer_manager/network_state/tier1.rs:124]: It is exposed on the debug page, as AccountData. It is not really feasible to expose it in prometheus (gauge cannot have a string value), unless we abuse the system.

File: chain/network/src/config_json.rs (hunk 4 above)
Comment (pompon0) [reply chain/network/src/config_json.rs:184]: done

**Pull request merged**